src/core/geometry/qgsnurbscurve.cpp

```diff
@@ -0,0 +1,1622 @@
+  if ( !o )
+    return false;
+
+  if ( mDegree != o->mDegree || mClosed != o->mClosed || mControlPoints.size() != o->mControlPoints.size() || mWeights.size() != o->mWeights.size() || mKnots.size() != o->mKnots.size() )
+  {
+    return false;
+  }
+
+  for ( int i = 0; i < mControlPoints.size(); ++i )
+  {
+    if ( mControlPoints[i].distance( o->mControlPoints[i] ) >= epsilon )
+      return false;
+  }
+
+  for ( int i = 0; i < mWeights.size(); ++i )
+  {
+    if ( std::fabs( mWeights[i] - o->mWeights[i] ) > epsilon )
+      return false;
+  }
+
+  for ( int i = 0; i < mKnots.size(); ++i )
+  {
+    if ( std::fabs( mKnots[i] - o->mKnots[i] ) > epsilon )
+      return false;
+  }
+
+  return true;
+}
+
+bool QgsNurbsCurve::fuzzyDistanceEqual( const QgsAbstractGeometry &other, double epsilon ) const
+{
+  const QgsNurbsCurve *o = qgsgeometry_cast<const QgsNurbsCurve *>( &other );
+  if ( !o )
+    return false;
+
+  if ( mDegree != o->mDegree || mClosed != o->mClosed || mControlPoints.size() != o->mControlPoints.size() || mWeights.size() != o->mWeights.size() || mKnots.size() != o->mKnots.size() )
+  {
+    return false;
+  }
+
+  for ( int i = 0; i < mControlPoints.size(); ++i )
+  {
+    if ( mControlPoints[i].distance( o->mControlPoints[i] ) > epsilon )
+      return false;
+  }
+
+  for ( int i = 0; i < mWeights.size(); ++i )
+  {
+    if ( std::fabs( mWeights[i] - o->mWeights[i] ) > epsilon )
+      return false;
+  }
+
+  for ( int i = 0; i < mKnots.size(); ++i )
+  {
+    if ( std::fabs( mKnots[i] - o->mKnots[i] ) > epsilon )
+      return false;
+  }
+
+  return true;
+}
+
+QString QgsNurbsCurve::geometryType() const
+{
+  return QStringLiteral( "NurbsCurve" );
+}
+
+bool QgsNurbsCurve::hasCurvedSegments() const
+{
+  return true;
+}
+
+int QgsNurbsCurve::partCount() const
+{
+  return 1;
+}
+
+QgsCurve *QgsNurbsCurve::toCurveType() const
+{
+  return new QgsNurbsCurve( *this );
+}
+
+QgsPoint QgsNurbsCurve::vertexAt( QgsVertexId id ) const
+{
+  if ( id.part != 0 || id.ring != 0 )
+  {
+    return QgsPoint();
+  }
+  const int idx = id.vertex;
+  if ( idx < 0 || idx >= mControlPoints.size() )
+  {
+    return QgsPoint();
+  }
+  return mControlPoints[idx];
+}
+
+int QgsNurbsCurve::vertexCount( int part, int ring ) const
+{
+  return ( part == 0 && ring == 0 ) ? mControlPoints.size() : 0;
+}
+
+int QgsNurbsCurve::vertexNumberFromVertexId( QgsVertexId id ) const
+{
+  if ( id.part == 0 && id.ring == 0 )
+  {
+    return id.vertex;
+  }
+  return -1;
+}
+
+bool QgsNurbsCurve::isValid( QString &error, Qgis::GeometryValidityFlags flags ) const
+{
+  Q_UNUSED( flags );
+
+  if ( mDegree < 1 )
+  {
+    error = QStringLiteral( "Degree must be >= 1" );
+    return false;
+  }
+
+  const int n = mControlPoints.size();
+  if ( n < mDegree + 1 )
+  {
+    error = QStringLiteral( "Not enough control points for degree" );
+    return false;
+  }
+
+  if ( mKnots.size() != n + mDegree + 1 )
```

Comment:
shouldn't we check that the value in mKnots are non-decreasing ? (otherwise findKnotSpan() could misbehave)